### ISSUE TITLE
Print buton does not work (print.js error)

### ISSUE THREAD

**Issue by scrmtrey91** [vaadin-component-factory/vcf-pdf-viewer-flow#24]
Print button is displayed correctly but it does not work, it does not print the PDF.

![image](https://user-images.githubusercontent.com/24773401/201517263-859a7d8d-0080-4ccd-858e-09a57710323d.png)

Error is in javascript of print.js:

```js
Uncaught (webpack-internal:///../target/flow-frontend/print-js/dist/print.js:468:9)
```

![image](https://user-images.githubusercontent.com/24773401/201517268-596be543-ccb4-4022-aa20-f17401bee08c.png)

I need this feature to work for clients. Now they download the PDF first then print it with windows or browser native viewer.

But i really want to use your addon because it is great and clean with only the necessary functions.

**Comment by paodb** [comment]
Can you share a reproducible example or more information about your scenario (vaadin version, component version, if this is browser specific, etc)? I did a quick test in the latest Vaadin version and latest component version and printing works okay. See 
[pdf-viewer-test.zip](https://github.com/vaadin-component-factory/vcf-pdf-viewer-flow/files/10006668/pdf-viewer-test.zip).

**Comment by scrmtrey91** [comment]
Your example works (printing). Hmm interesing, maybe it works because of springboot. I am not using springboot.

I am using java 17, pnpm, webpack latest vaadin and latest addon version. The app is deployed on Payara 5.2022.3.

Sorry i can not share the real app with you.

**Comment by scrmtrey91** [comment]
New error:
```js
Uncaught Not Found (webpack-internal:///../target/flow-frontend/print-js/dist/print.js:468:9)
```

**Comment by scrmtrey91** [comment]
I looked into developer tools , console in chrome...and it looks like the PDF cannot be fetch when the print button is clicked. 

The reason why print.js does not work (no data pdf to print).

![image](https://user-images.githubusercontent.com/24773401/202866187-0b408b05-ca52-4dbb-a9c0-0a71b027b035.png)

**Comment by scrmtrey91** [comment]
It seems vaadin build wrong path to dynamic content. It is missing the app context root before /VAADIN/...

**Comment by scrmtrey91** [comment]
Download button is using correct URL path to dynamic content : 

![image](https://user-images.githubusercontent.com/24773401/202867419-6a3ed0d4-c820-439c-b47d-0057dafb9e39.png)

It is using context root path befoe /VAADIN/ , thats why resource is resolved and download works.

**Comment by scrmtrey91** [comment]
Steps to reproduce error:



_1) Go to : https://vaadin.com/hello-world-starters_

_2) Choose started : VAADIN 23 - CDI_

_3) add pdf addon + pdf files in resources + web pack flag_

_4) Run payara , deploy war_

_5) Print PDF NOT working, download working OK_


**Zip project sample not working properly:** 
[skeleton-starter-flow-cdi-23.zip](https://github.com/vaadin-component-factory/vcf-pdf-viewer-flow/files/10047809/skeleton-starter-flow-cdi-23.zip)

**Comment by ARCHERS** [comment]
I have a similar issue with both the download and print function. I'm using Spring boot, Vaadin 23 and java 17. Rather than reading a pdf file from disc as in the pdf-viewer-test sample provided above, I'm using JasperExportManager to export to a temporary pdf file:

jasperPrint = JasperFillManager.fillReport(filePath, param, conUser);
 reportFile  = File.createTempFile("output", ".pdf");
JasperExportManager.exportReportToPdfFile(jasperPrint, reportFile.getAbsolutePath());
fs = getInputStream(reportFile);
resource = new StreamResource(displayName, () -> fs);
pdfViewer = new PdfViewer();
pdfViewer.setSrc(resource);

**Comment by scrmtrey91** [comment]
any news about that ?

**Comment by scrmtrey91** [comment]
Any fixes for print button not working ? 

It takes wrong URL for file (without context root of app) thus file not found 404 and print button not working.

**Comment by scrmtrey91** [comment]
**The problem is in file "print.js" in line 811:**

```javascript
params.printable = /^(blob|http|\/\/)/i.test(params.printable) ? params.printable : window.location.origin + (params.printable.charAt(0) !== '/' ? '/' + params.printable : params.printable); // Get the file through a http request (Preload)
```

It only gets origin of window which does not include app context root.

If the app is deployed on server behind reverse-proxy (the url of app is mapped by external domain, context path is hidden), then the print works.

**Example:** 
1) https://localhost:8181/AppRoot  -> print not working (it only gets https://localhost:8181/)
2) https://app-root-proxy/ -> print works (correct origin because of no port and context root: https://app-root-proxy/)

**Comment by renojjoseph** [comment]
I also have the same poblem  Download button and Print button not working.
Using Vaadin 24, Sprig boot 3.0.1 and vcf-pdf-viewer 2.7.1

Case 1. (As in the example)
` StreamResource resource = new StreamResource("example.pdf", () -> getClass().getResourceAsStream("/pdf/example.pdf"));`  // shows pdf and downloadable also but print button not working.  in this case pdf is present in resouces floder.

Case 2. (As commented by [ARCHERS](https://github.com/ARCHERS))
 ```
JasperPrint jp = JasperFillManager.fillReport(jasper report, params, con);
 File reportFile  = File.createTempFile("output", ".pdf");
  JasperExportManager.exportReportToPdfFile(jp, reportFile.getAbsolutePath());
 InputStream targetStream = new FileInputStream(reportFile.getAbsolutePath());
 StreamResource resource = new StreamResource("example.pdf", () -> targetStream);
```
// Shows pdf but both download and print button not working. pdf is not present in resouces floder. making only temp file

Case3. byte[] pdfBytes = JasperRunManager.runReportToPdf(jasper report, params, con);
StreamResource resource = new StreamResource("example.pdf",() -> getClass().getResourceAsStream(pdfBytes));
// Shows pdf but both download and print button not working. pdf is not present in resouces floder. making resource from byteArray.

it is downloadable/printable only if a pdf is present in resourse folder.

**Comment by ARCHERS** [comment]
I now have the download and print buttons working with an inputstream source using the following code:
```
            jasperPrint = JasperFillManager.fillReport(filePath, param, conUser);
            byte[] byteArray = JasperExportManager.exportReportToPdf(jasperPrint);
            resource = new StreamResource(displayName, () -> new ByteArrayInputStream(byteArray));
            resource.setContentType("application/pdf");
            pdfViewer.setSrc(resource);
```
I had originally created a temp file from JasperPrint, but here the StreamResource is derived from a PDF byte array by way of ByteArrayInputStream. I'm using pdfviewer version 2.7.2 in Vaadin 23.3.5.

**Comment by scrmtrey91** [comment]
I have exact same code as you (from 1 year ago), always from byte array input stream (it is also jasper report byte pdf data).

BUT still does not work. Download working, print not working. 

pdfviewer version 2.7.2 in Vaadin 24.0.4 (but even with vaadin 23 does not work).

**Comment by ARCHERS** [comment]
So I'm running my code in development with no reverse proxy at http://localhost:8080/ - not tried it in production server yet. My issue (see previous post) was that when the InputStream is assigned to a variable before instantiating StreamResource, the PDF is displayed correctly in the viewer, but the download creates an empty PDF and the print does nothing because the stream is empty (had been consumed, but I don't get the print.js javascript error you described above). If the StreamResource is instantiated directly with "new ByteArrayInputStream(byteArray"), the PDF is populated correctly.

So this produces an empty (corrupt) PDF:
```
InputStream is = new ByteArrayInputStream(byteArray);
resource = new StreamResource(displayName, () -> is);
```
But this works:

`resource = new StreamResource(displayName, () -> new ByteArrayInputStream(byteArray));`

However I think that may be different to your problem.

**Comment by scrmtrey91** [comment]
Try running the app with context root instead. 

Example: /AppName

https/localhost:8181/AppName

Without context root (empty) it is working.

**Comment by ARCHERS** [comment]
You're right, I can reproduce the error by configuring the root context path. As you said above, it seems to be a problem with "window.location origin" in print,js which only gets the base URL. I think it should be window.location.pathname as per this:

https://www.samanthaming.com/tidbits/86-window-location-cheatsheet/

**Comment by scrmtrey91** [comment]
Yp. Probably fix. I guess it will never be fixed.

**Comment by TatuLund** [comment]
@scrmtrey91 Are you customer of ours with some subscription tier? If yes, which one?

**Comment by scrmtrey91** [comment]
We have Vaadin Pro Subscription.

**Comment by jm2rodriguez** [comment]
Hi,
I'm facing the same problem.
Finally I solve it by calling PdfViewer.setSrc() with a full url instead of the relative url generated by StreamResource.
It works fine for me.

// Create your StreamResource
StreamResource streamResource = new StreamResource(...);

StreamRegistration streamRegistration = VaadinSession.getCurrent().getResourceRegistry().registerResource(streamResource);

// Create a full string URI
...
String uri = scheme + "://" + serverName + ":" + serverPort + contextPath + "/" + streamRegistration.getResourceUri();

// Create a PdfViewer instance and set the source with the full URI
PdfViewer pdfViewer = new PdfViewer();
pdfViewer.setSrc(uri);
...

**Comment by scrmtrey91** [comment]
Need to replace this in print.js file :


`js
var printable = params.printable || '';
                        params.printable = /^(blob:|http?:|\/\/)/i.test(printable)
                            ? printable
                            : new URL(printable.replace(/^\/+/, ''), window.location.href).href;
`

- tested and it works
